docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+===========================================
+RFC 001: Storage Engines and Database Layer
+===========================================
+
+Changelog
+---------
+
+- 2021-04-19: Initial Draft (gist)
+- 2021-09-02: Migrated to RFC folder, with some updates  
+
+Abstract
+--------
+
+The aspect of Tendermint that's responsible for persistence and storage (often
+"the database" internally) represents a bottle neck in the architecture of the
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
+storage layer is handled generically, and most tests run with an in-memory
+framework, it's difficult to take advantage of any higher-level features of a
+database engine.
+
+Ideally, developers within Tendermint will be able to interact with persisted
+data via a single interface that can function, approximately like an object
```

Comment:
I'm not sure a _single_ interface is essential—if we have different data storage requirements (e.g., for logs vs. chain state vs. other things) I think it would be fine to define different interfaces for those. 

I don't know whether we actually have multiple storage concerns, but I do think it's essential that whatever interface(s) we _do_ define reflect the semantics of the consumer, not the provider. That is: The interface should express the domain requirements, not the implementation constraints.

(To be clear: I think this is more or less what you're advocating, not a contrast. And, of course, we have to choose requirements that _can_ be satisfied, but that's achievable)

Reply:
yeah, I think we're in agreement here. I deleted the word single.